File: test/jvm/com/cattailsw/nanidroid/install/NarArchiveInventoryValidatorTest.java
package com.cattailsw.nanidroid.install;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.assertArrayEquals;

import java.nio.charset.Charset;
import java.text.Normalizer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

import org.junit.Test;

public final class NarArchiveInventoryValidatorTest {
    private static final Charset UTF_8 = Charset.forName("UTF-8");
    private static final long MIB = 1024L * 1024L;

    @Test
    public void kotlinInventoryResultKeepsJavaDiagnosticFactorySemantics() {
        NarArchiveInventoryResult failure = NarArchiveInventoryResult.failure(
                NarInstallError.INVALID_ENTRY_METADATA, "central getter");

        assertFalse(failure.isSuccess());
        assertNull(failure.getInventory());
        assertEquals(NarInstallError.INVALID_ENTRY_METADATA, failure.getError());
        assertEquals("central getter", failure.getDetail());
    }
    private static final long FUZZ_SEED = 0x4e415244396231L;

    @Test
    public void inventoriesRootLayoutAndPreservesCentralIdentity() {
        Record descriptor = record("install.txt");
        Record payload = record("ghost/master/file.bin");
        payload.crc = 0x1234abcdL;
        payload.method = 8;
        payload.size = 321;
        payload.compressedSize = 123;

        NarArchiveInventoryResult result = validate(descriptor, payload);

        assertTrue(result.isSuccess());
        NarArchiveInventory inventory = result.getInventory();
        assertNull(inventory.getWrapperDirectory());
        assertEquals(0, inventory.getDescriptorOrdinal());
        assertEquals(2, inventory.getEntries().size());
        NarArchiveInventory.Entry planned = inventory.getEntries().get(1);
        assertEquals(1, planned.getOrdinal());
        assertEquals(payload.rawName, planned.getRawName());
        assertEquals(payload.rawName, planned.getNormalizedArchivePath());
        assertEquals(payload.rawName, planned.getRelativePath());
        assertEquals(payload.crc, planned.getCrc());
        assertEquals(payload.method, planned.getMethod());
        assertEquals(payload.size, planned.getDeclaredSize());
        assertEquals(payload.compressedSize, planned.getCompressedSize());
        try {
            inventory.getEntries().clear();
            throw new AssertionError("inventory entries must be immutable");
        } catch (UnsupportedOperationException expected) {
            // Expected.
        }
    }

    @Test
    public void stripsOneUniformWrapperButRetainsEveryRawEntry() {
        NarArchiveInventoryResult result = validate(
                directory("bundle/"),
                record("bundle/install.txt"),
                record("bundle/ghost/master/file.txt"));

        assertTrue(result.isSuccess());
        NarArchiveInventory inventory = result.getInventory();
        assertEquals("bundle", inventory.getWrapperDirectory());
        assertEquals(1, inventory.getDescriptorOrdinal());
        assertEquals(
                Arrays.asList(null, "install.txt", "ghost/master/file.txt"),
                relativePaths(inventory));
        assertEquals("bundle/", inventory.getEntries().get(0).getRawName());
        assertFalse(inventory.getEntries().get(0).isInstallEntry());
        assertTrue(inventory.getEntries().get(1).isInstallEntry());
    }

    @Test
    public void rejectsUnsupportedDescriptorLayouts() {
        rejectsNames(NarInstallError.MISSING_INSTALL_DESCRIPTOR, "readme.txt");
        rejectsNames(NarInstallError.MISSING_INSTALL_DESCRIPTOR, "INSTALL.TXT");
        rejectsNames(
                NarInstallError.AMBIGUOUS_LAYOUT,
                "install.txt", "bundle/install.txt");
        rejectsNames(
                NarInstallError.AMBIGUOUS_LAYOUT,
                "one/install.txt", "two/install.txt");
        rejectsNames(
                NarInstallError.MIXED_LAYOUT,
                "bundle/install.txt", "sibling.txt");
        rejectsNames(NarInstallError.INVALID_LAYOUT, "one/two/install.txt");
        rejectsNames(
                NarInstallError.INVALID_LAYOUT,
                "install.txt", "one/two/install.txt");
        rejectsNames(
                NarInstallError.MISSING_INSTALL_DESCRIPTOR, "install.txt/");
    }

    @Test
    public void rejectsHostilePaths() {
        String[] hostile = {
            "../escape",
            "/absolute",
            "nested\\file",
            "C:drive",
            "\\\\server\\share",
            "bad\u0000name",
            "bad\u0001name",
            "./file",
            "one/../file",
            "one//file",
            "bad\uD800name",
        };
        for (String path : hostile) {
            rejectsNames(NarInstallError.INVALID_PATH, "install.txt", path);
        }
    }

    @Test
    public void distinguishesDuplicatesNormalizationAndImplicitCollisions() {
        rejectsNames(
                NarInstallError.DUPLICATE_ENTRY,
                "install.txt", "ghost/file", "ghost/file");
        rejectsNames(
                NarInstallError.NORMALIZED_COLLISION,
                "install.txt", "Ghost/File", "ghost/file");

        String composed = Normalizer.normalize("caf\u00e9", Normalizer.Form.NFC);
        String decomposed = Normalizer.normalize(composed, Normalizer.Form.NFD);
        rejectsNames(
                NarInstallError.NORMALIZED_COLLISION,
                "install.txt", composed + "/one", decomposed + "/two");
        rejectsNames(
                NarInstallError.NORMALIZED_COLLISION,
                "install.txt", "Ghost/one", "ghost/two");
    }

    @Test
    public void lowerCaseThenNfcCollisionIsRejectedInBothOrders() {
        String decomposedAfterLowerCase = "J\u030C";
        String composedLowerCase = "\u01F0";
        String[][] orders = {
            {decomposedAfterLowerCase, composedLowerCase},
            {composedLowerCase, decomposedAfterLowerCase},
        };
        for (String[] order : orders) {
            rejectsNames(
                    NarInstallError.NORMALIZED_COLLISION,
                    "install.txt", order[0], order[1]);
            rejectsNames(
                    NarInstallError.NORMALIZED_COLLISION,
                    "install.txt", order[0] + "/one", order[1] + "/two");
        }
    }

    @Test
    public void snapshotsEachCentralGetterExactlyOnce() {
        OneShotEntry descriptor = new OneShotEntry(-1);

        NarArchiveInventoryResult result =
                new NarArchiveInventoryValidator().validate(
                        Arrays.asList(descriptor));

        assertTrue(result.isSuccess());
        assertArrayEquals(new int[] {1, 1, 1, 1, 1, 1, 1}, descriptor.reads);
        NarArchiveInventory.Entry entry = result.getInventory().getEntries().get(0);
        assertEquals("install.txt", entry.getRawName());
        assertEquals(0x1234L, entry.getCrc());
        assertEquals(8, entry.getMethod());
        assertEquals(10, entry.getDeclaredSize());
        assertEquals(6, entry.getCompressedSize());
    }

    @Test
    public void mapsCentralGetterRuntimeFailureToTypedError() {
        NarArchiveInventoryResult result =
                new NarArchiveInventoryValidator().validate(
                        Arrays.asList(new OneShotEntry(1)));

        assertError(NarInstallError.INVALID_ENTRY_METADATA, result);
    }

    @Test
    public void mapsRawNfdNameToNfcOutputWithoutLosingIdentity() {
        String normalized = "ghost/"
                + Normalizer.normalize("caf\u00e9.txt", Normalizer.Form.NFC);
        String raw = Normalizer.normalize(normalized, Normalizer.Form.NFD);

        NarArchiveInventoryResult result =
                validate(record("install.txt"), record(raw));

        assertTrue(result.isSuccess());
        NarArchiveInventory.Entry entry = result.getInventory().getEntries().get(1);
        assertEquals(raw, entry.getRawName());
        assertEquals(normalized, entry.getNormalizedArchivePath());
        assertEquals(normalized, entry.getRelativePath());
    }

    @Test
    public void rejectsFileDirectoryCollisionsInEveryOrdering() {
        rejectsNames(
                NarInstallError.FILE_DIRECTORY_COLLISION,
                "install.txt", "ghost", "ghost/master/file");
        rejectsNames(
                NarInstallError.FILE_DIRECTORY_COLLISION,
                "install.txt", "ghost/master/file", "ghost");
        rejectsNames(
                NarInstallError.FILE_DIRECTORY_COLLISION,
                "install.txt", "ghost", "ghost/");
    }

    @Test
    public void enforcesExactCountDepthPathComponentAndRawNameBoundaries() {
        List<Record> entries = new ArrayList<Record>();
        entries.add(record("install.txt"));
        for (int index = 1; index < 10000; index++) {
            entries.add(record("payload/" + index));
        }
        assertTrue(validate(entries).isSuccess());
        entries.add(record("payload/overflow"));
        assertError(NarInstallError.ENTRY_COUNT_LIMIT, validate(entries));

        assertTrue(validateWithPath(repeatedPath(32, 1)).isSuccess());
        rejectsPath(NarInstallError.PATH_DEPTH_LIMIT, repeatedPath(33, 1));
        assertTrue(validateWithPath(asciiPath(1024)).isSuccess());
        rejectsPath(NarInstallError.PATH_LENGTH_LIMIT, asciiPath(1025));
        assertTrue(validateWithPath(repeat('a', 255)).isSuccess());
        rejectsPath(NarInstallError.COMPONENT_LENGTH_LIMIT, repeat('a', 256));

        String malformedAtLimit = repeat('a', 4095) + "\uD800";
        rejectsPath(NarInstallError.INVALID_PATH, malformedAtLimit);
        rejectsPath(NarInstallError.RAW_NAME_LENGTH_LIMIT, malformedAtLimit + "a");
    }

    @Test
    public void enforcesDescriptorAndDeclaredSizeRatioBoundaries() {
        Record descriptor = record("install.txt");
        descriptor.size = 64 * 1024;
        descriptor.compressedSize = -1;
        assertTrue(validate(descriptor).isSuccess());
        descriptor.size++;
        rejects(NarInstallError.INSTALL_DESCRIPTOR_LIMIT, descriptor);

        descriptor = record("install.txt");
        Record payload = sized("payload", 128 * MIB, -1);
        assertTrue(validate(descriptor, payload).isSuccess());
        payload.size++;
        rejects(NarInstallError.DECLARED_ENTRY_SIZE_LIMIT, descriptor, payload);

        List<Record> total = new ArrayList<Record>();
        total.add(descriptor);
        for (int index = 0; index < 4; index++) {
            total.add(sized("part-" + index, 128 * MIB, -1));
        }
        assertTrue(validate(total).isSuccess());
        total.add(sized("overflow", 1, -1));
        assertError(NarInstallError.DECLARED_TOTAL_SIZE_LIMIT, validate(total));

        assertTrue(validate(descriptor, sized("ratio", 1000, 1)).isSuccess());
        rejects(
                NarInstallError.DECLARED_RATIO_LIMIT,
                descriptor, sized("ratio", 1001, 1));
        rejects(
                NarInstallError.DECLARED_RATIO_LIMIT,
                descriptor, sized("ratio", 1, 0));
        assertTrue(validate(descriptor, sized("unknown", -1, -1)).isSuccess());
    }

    @Test
    public void rejectsInvalidCentralRecordsAndAllowsDocumentedUnknowns() {
        List<Record> nullRecord = new ArrayList<Record>();
        nullRecord.add(null);
        assertError(
                NarInstallError.INVALID_ENTRY_METADATA,
                validateWithoutOrdinalAssignment(nullRecord));
        rejects(NarInstallError.INVALID_PATH, new Record(null, false));
        rejects(NarInstallError.INVALID_PATH, record(""));
        rejects(NarInstallError.INVALID_ENTRY_METADATA, new Record("file/", false));
        rejects(
                NarInstallError.INVALID_ENTRY_METADATA,
                new Record("directory", true));

        Record descriptor = record("install.txt");
        assertError(
                NarInstallError.INVALID_ENTRY_METADATA,
                validateWithoutOrdinalAssignment(descriptor));
        descriptor.ordinal = 7;
        assertError(
                NarInstallError.INVALID_ENTRY_METADATA,
                validateWithoutOrdinalAssignment(descriptor));
        Record gap = record("payload");
        descriptor.ordinal = 0;
        gap.ordinal = 2;
        assertError(
                NarInstallError.INVALID_ENTRY_METADATA,
                validateWithoutOrdinalAssignment(descriptor, gap));
        gap.ordinal = 0;
        assertError(
                NarInstallError.INVALID_ENTRY_METADATA,
                validateWithoutOrdinalAssignment(descriptor, gap));

        for (int field = 0; field < 6; field++) {
            Record invalid = record("install.txt");
            invalid.ordinal = 0;
            if (field == 0) invalid.size = -2;
            if (field == 1) invalid.compressedSize = -2;
            if (field == 2) invalid.crc = -2;
            if (field == 3) invalid.crc = 0x100000000L;
            if (field == 4) invalid.method = -2;
            if (field == 5) invalid.method = 7;
            assertError(
                    NarInstallError.INVALID_ENTRY_METADATA,
                    validateWithoutOrdinalAssignment(invalid));
        }
        Record unknown = sized("install.txt", -1, -1);
        unknown.crc = -1;
        unknown.method = -1;
        NarArchiveInventoryResult unknownResult = validate(unknown);
        assertTrue(unknownResult.isSuccess());
        assertEquals(-1, unknownResult.getInventory().getDeclaredTotalSize());
    }

    @Test
    public void deterministicFuzzMutatesValidPathsWithExactErrors() {
        Random random = new Random(FUZZ_SEED);
        for (int index = 0; index < 512; index++) {
            String baseline = "root" + random.nextInt(100000)
                    + "/leaf" + index + ".txt";
            assertTrue(validateWithPath(baseline).isSuccess());
            String[] mutations = {
                "../" + baseline,
                "/" + baseline,
                baseline.replace("/", "//"),
                baseline.replace("/", "\\"),
                baseline + "\u0001",
                baseline.substring(0, baseline.indexOf('/') + 1) + repeat('a', 256),
                repeatedPath(31, 1) + "/" + baseline,
                baseline + "/" + asciiPath(
                        1025 - baseline.getBytes(UTF_8).length - 1),
                baseline + "/" + repeat('a', 4097 - baseline.length() - 1),
                baseline + "\uD800",
            };
            NarInstallError[] errors = {
                NarInstallError.INVALID_PATH,
                NarInstallError.INVALID_PATH,
                NarInstallError.INVALID_PATH,
                NarInstallError.INVALID_PATH,
                NarInstallError.INVALID_PATH,
                NarInstallError.COMPONENT_LENGTH_LIMIT,
                NarInstallError.PATH_DEPTH_LIMIT,
                NarInstallError.PATH_LENGTH_LIMIT,
                NarInstallError.RAW_NAME_LENGTH_LIMIT,
                NarInstallError.INVALID_PATH,
            };
            int mutation = index % mutations.length;
            assertError(errors[mutation], validateWithPath(mutations[mutation]));
        }
    }

    private static NarArchiveInventoryResult validate(Record... records) {
        return validate(new ArrayList<Record>(Arrays.asList(records)));
    }

    private static NarArchiveInventoryResult validate(List<Record> records) {
        for (int index = 0; index < records.size(); index++) {
            if (records.get(index).ordinal < 0) {
                records.get(index).ordinal = index;
            }
        }
        return validateWithoutOrdinalAssignment(records);
    }

    private static NarArchiveInventoryResult validateWithoutOrdinalAssignment(
            Record... records) {
        return validateWithoutOrdinalAssignment(Arrays.asList(records));
    }

    private static NarArchiveInventoryResult validateWithoutOrdinalAssignment(
            List<Record> records) {
        return new NarArchiveInventoryValidator().validate(records);
    }

    private static NarArchiveInventoryResult validateWithPath(String path) {
        return validate(record("install.txt"), record(path));
    }

    private static void rejects(NarInstallError error, Record... records) {
        assertError(error, validate(records));
    }

    private static void rejectsNames(NarInstallError error, String... names) {
        Record[] records = new Record[names.length];
        for (int index = 0; index < names.length; index++) {
            records[index] = record(names[index]);
        }
        rejects(error, records);
    }

    private static void rejectsPath(NarInstallError error, String path) {
        assertError(error, validateWithPath(path));
    }

    private static void assertError(
            NarInstallError expected,
            NarArchiveInventoryResult result) {
        assertFalse(result.isSuccess());
        assertEquals(expected, result.getError());
        assertNull(result.getInventory());
    }

    private static List<String> relativePaths(NarArchiveInventory inventory) {
        List<String> paths = new ArrayList<String>();
        for (NarArchiveInventory.Entry entry : inventory.getEntries()) {
            paths.add(entry.getRelativePath());
        }
        return paths;
    }

    private static Record record(String name) {
        return new Record(name, name.endsWith("/"));
    }

    private static Record directory(String name) {
        return new Record(name, true);
    }

    private static Record sized(String name, long size, long compressedSize) {
        Record result = record(name);
        result.size = size;
        result.compressedSize = compressedSize;
        return result;
    }

    private static String repeatedPath(int count, int componentLength) {
        StringBuilder result = new StringBuilder();
        for (int index = 0; index < count; index++) {
            if (index > 0) {
                result.append('/');
            }
            result.append(repeat('a', componentLength));
        }
        return result.toString();
    }

    private static String asciiPath(int byteCount) {
        int letters = byteCount - 4;
        int base = letters / 5;
        int remainder = letters % 5;
        StringBuilder result = new StringBuilder(byteCount);
        for (int index = 0; index < 5; index++) {
            if (index > 0) {
                result.append('/');
            }
            result.append(repeat('a', base + (index < remainder ? 1 : 0)));
        }
        assertEquals(byteCount, result.toString().getBytes(UTF_8).length);
        return result.toString();
    }

    private static String repeat(char value, int count) {
        char[] content = new char[count];
        Arrays.fill(content, value);
        return new String(content);
    }

    private static final class Record
            implements NarArchiveInventoryValidator.CentralEntry {
        private int ordinal = -1;
        private final String rawName;
        private final boolean directory;
        private long crc = -1;
        private int method = -1;
        private long size = 0;
        private long compressedSize = 0;

        private Record(String rawName, boolean directory) {
            this.rawName = rawName;
            this.directory = directory;
        }

        @Override public int getOrdinal() { return ordinal; }
        @Override public String getRawName() { return rawName; }
        @Override public boolean isDirectory() { return directory; }
        @Override public long getCrc() { return crc; }
        @Override public int getMethod() { return method; }
        @Override public long getDeclaredSize() { return size; }
        @Override public long getCompressedSize() { return compressedSize; }
    }

    private static final class OneShotEntry
            implements NarArchiveInventoryValidator.CentralEntry {
        private final int throwOnFirst;
        private final int[] reads = new int[7];

        private OneShotEntry(int throwOnFirst) {
            this.throwOnFirst = throwOnFirst;
        }

        private Object read(int field, Object value) {
            reads[field]++;
            if (field == throwOnFirst || reads[field] != 1) {
                throw new IllegalStateException("getter " + field);
            }
            return value;
        }

        @Override public int getOrdinal() {
            return ((Integer) read(0, Integer.valueOf(0))).intValue();
        }
        @Override public String getRawName() {
            return (String) read(1, "install.txt");
        }
        @Override public boolean isDirectory() {
            return ((Boolean) read(2, Boolean.FALSE)).booleanValue();
        }
        @Override public long getCrc() {
            return ((Long) read(3, Long.valueOf(0x1234L))).longValue();
        }
        @Override public int getMethod() {
            return ((Integer) read(4, Integer.valueOf(8))).intValue();
        }
        @Override public long getDeclaredSize() {
            return ((Long) read(5, Long.valueOf(10))).longValue();
        }
        @Override public long getCompressedSize() {
            return ((Long) read(6, Long.valueOf(6))).longValue();
        }
    }
}

File: test/jvm/com/cattailsw/nanidroid/install/NarArchiveInventoryValidatorTest.kt
package com.cattailsw.nanidroid.install

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/** Kotlin port of central-directory inventory behavior characterization. */
class NarArchiveInventoryValidatorTest {
    @Test fun acceptsFlatAndWrappedLayoutsWithStableCentralMapping() {
        val flat = validate(file("install.txt", 20), file("ghost/master/surface0.png", 4))
        assertTrue(flat.isSuccess()); assertNull(flat.getInventory()!!.getWrapperDirectory()); assertEquals(0, flat.getInventory()!!.getDescriptorOrdinal())
        assertEquals(listOf("install.txt", "ghost/master/surface0.png"), flat.getInventory()!!.getEntries().map { it.getRelativePath() })
        val wrapped = validate(dir("wrapper"), file("wrapper/install.txt", 20), file("wrapper/ghost/master/surface0.png", 4))
        assertTrue(wrapped.isSuccess()); assertEquals("wrapper", wrapped.getInventory()!!.getWrapperDirectory())
        assertEquals(listOf(null, "install.txt", "ghost/master/surface0.png"), wrapped.getInventory()!!.getEntries().map { it.getRelativePath() })
    }

    @Test fun kotlinInventoryResultKeepsJavaDiagnosticFactorySemantics() {
        val failure = NarArchiveInventoryResult.failure(NarInstallError.INVALID_ENTRY_METADATA, "central getter")
        assertFalse(failure.isSuccess()); assertNull(failure.getInventory()); assertEquals(NarInstallError.INVALID_ENTRY_METADATA, failure.getError()); assertEquals("central getter", failure.getDetail())
    }

    @Test fun preservesExactCentralIdentityAndNfcOutput() {
        val payload = Record("ghost/master/file.bin", false, crcValue = 0x1234abcdL, methodValue = 8, sizeValue = 321, compressedValue = 123)
        val result = validate(file("install.txt", 1), payload); val entry = result.getInventory()!!.getEntries()[1]
        assertEquals(1, entry.getOrdinal()); assertEquals(payload.raw, entry.getRawName()); assertEquals(payload.raw, entry.getNormalizedArchivePath()); assertEquals(payload.raw, entry.getRelativePath()); assertEquals(0x1234abcdL, entry.getCrc()); assertEquals(321, entry.getDeclaredSize()); assertEquals(123, entry.getCompressedSize())
        val raw = java.text.Normalizer.normalize("ghost/café.txt", java.text.Normalizer.Form.NFD)
        val nfc = java.text.Normalizer.normalize(raw, java.text.Normalizer.Form.NFC)
        val normalized = validate(file("install.txt", 1), file(raw!!, 1)).getInventory()!!.getEntries()[1]
        assertEquals(raw, normalized.getRawName()); assertEquals(nfc, normalized.getNormalizedArchivePath())
    }

    @Test fun rejectsCentralMetadataCollisionsAndUnsafePaths() {
        assertError(NarInstallError.INVALID_ENTRY_METADATA, NarArchiveInventoryValidator().validate(null))
        assertError(NarInstallError.INVALID_PATH, validate(file("../install.txt", 1)))
        assertError(NarInstallError.DUPLICATE_ENTRY, validate(file("install.txt", 1), file("install.txt", 1)))
        assertError(NarInstallError.FILE_DIRECTORY_COLLISION, validate(file("ghost", 1), file("ghost/install.txt", 1)))
        assertError(NarInstallError.NORMALIZED_COLLISION, validate(file("install.txt", 1), file("INSTALL.TXT", 1)))
        assertError(NarInstallError.MISSING_INSTALL_DESCRIPTOR, validate(file("payload", 1)))
        assertError(NarInstallError.INVALID_LAYOUT, validate(file("a/b/install.txt", 1)))
        assertError(NarInstallError.AMBIGUOUS_LAYOUT, validate(file("install.txt", 1), file("wrapper/install.txt", 1)))
        assertError(NarInstallError.MIXED_LAYOUT, validate(file("wrapper/install.txt", 1), file("outside", 1)))
    }

    @Test fun enforcesCountDepthLengthComponentAndRawNameBoundaries() {
        val entries = mutableListOf(file("install.txt", 1)); repeat(9_999) { entries += file("payload/$it", 1) }
        assertTrue(validate(entries).isSuccess()); entries += file("payload/overflow", 1); assertError(NarInstallError.ENTRY_COUNT_LIMIT, validate(entries))
        assertTrue(validateWithPath(repeatedPath(32, 1)).isSuccess()); assertError(NarInstallError.PATH_DEPTH_LIMIT, validateWithPath(repeatedPath(33, 1)))
        assertTrue(validateWithPath(asciiPath(1024)).isSuccess()); assertError(NarInstallError.PATH_LENGTH_LIMIT, validateWithPath(asciiPath(1025)))
        assertTrue(validateWithPath("a".repeat(255)).isSuccess()); assertError(NarInstallError.COMPONENT_LENGTH_LIMIT, validateWithPath("a".repeat(256)))
        val malformed = "a".repeat(4095) + '\uD800'; assertError(NarInstallError.INVALID_PATH, validateWithPath(malformed)); assertError(NarInstallError.RAW_NAME_LENGTH_LIMIT, validateWithPath(malformed + "a"))
    }

    @Test fun rejectsNfcAndCaseCollisionsInBothOrders() {
        assertError(NarInstallError.NORMALIZED_COLLISION, validate(file("install.txt", 1), file("Ghost/File", 1), file("ghost/file", 1)))
        val a = "J\u030C"; val b = "ǰ"
        listOf(a to b, b to a).forEach { (first, second) ->
            assertError(NarInstallError.NORMALIZED_COLLISION, validate(file("install.txt", 1), file(first, 1), file(second, 1)))
            assertError(NarInstallError.NORMALIZED_COLLISION, validate(file("install.txt", 1), file("$first/one", 1), file("$second/two", 1)))
        }
    }

    @Test fun enforcesDeclaredSizesCompressionAndDescriptorLimits() {
        assertError(NarInstallError.INSTALL_DESCRIPTOR_LIMIT, validate(file("install.txt", 64L * 1024 + 1)))
        assertError(NarInstallError.DECLARED_ENTRY_SIZE_LIMIT, validate(file("install.txt", 1), file("payload", 128L * 1024 * 1024 + 1)))
        assertError(NarInstallError.DECLARED_RATIO_LIMIT, validate(Record("install.txt", false, sizeValue = 1001, compressedValue = 1)))
        assertError(NarInstallError.INVALID_ENTRY_METADATA, validate(Record("install.txt", false, crcValue = -2)))
    }

    @Test fun enforcesTotalAndExactCompressionBoundaries() {
        val descriptor = Record("install.txt", false, sizeValue = 0, compressedValue = -1); val total = mutableListOf(descriptor); repeat(4) { total += Record("part-$it", false, sizeValue = 128L * 1024 * 1024, compressedValue = -1) }
        assertTrue(validate(total).isSuccess()); total += Record("overflow", false, sizeValue = 1, compressedValue = -1); assertError(NarInstallError.DECLARED_TOTAL_SIZE_LIMIT, validate(total))
        assertTrue(validate(descriptor, Record("ratio", false, sizeValue = 1000, compressedValue = 1)).isSuccess())
        assertError(NarInstallError.DECLARED_RATIO_LIMIT, validate(descriptor, Record("ratio", false, sizeValue = 1001, compressedValue = 1)))
        assertError(NarInstallError.DECLARED_RATIO_LIMIT, validate(descriptor, Record("ratio", false, sizeValue = 1, compressedValue = 0)))
        assertTrue(validate(descriptor, Record("unknown", false, sizeValue = -1, compressedValue = -1)).isSuccess())
    }

    @Test fun mapsGetterFailuresAndRejectsInvalidCentralMetadata() {
        assertError(NarInstallError.INVALID_ENTRY_METADATA, NarArchiveInventoryValidator().validate(listOf(ThrowingEntry())))
        assertError(NarInstallError.INVALID_ENTRY_METADATA, NarArchiveInventoryValidator().validate(listOf(null)))
        assertError(NarInstallError.INVALID_PATH, validate(Record(null, false)))
        assertError(NarInstallError.INVALID_PATH, validate(Record("", false)))
        assertError(NarInstallError.INVALID_ENTRY_METADATA, validate(Record("file/", false)))
        assertError(NarInstallError.INVALID_ENTRY_METADATA, validate(Record("directory", true)))
        listOf(
            Record("install.txt", false, sizeValue = -2), Record("install.txt", false, compressedValue = -2), Record("install.txt", false, crcValue = -2), Record("install.txt", false, crcValue = 0x100000000L), Record("install.txt", false, methodValue = -2), Record("install.txt", false, methodValue = 7)
        ).forEach { assertError(NarInstallError.INVALID_ENTRY_METADATA, validate(it)) }
        assertTrue(validate(Record("install.txt", false, crcValue = -1, methodValue = -1, sizeValue = -1, compressedValue = -1)).isSuccess())
    }

    @Test fun deterministicFuzzMutatesValidPathsWithExactErrors() {
        val random = java.util.Random(0x4e415244396231L)
        repeat(512) { index ->
            val baseline = "root${random.nextInt(100000)}/leaf$index.txt"
            assertTrue(validateWithPath(baseline).isSuccess())
            val mutations = listOf(
                "../$baseline", "/$baseline", baseline.replace("/", "//"), baseline.replace("/", "\\"), baseline + '\u0001',
                baseline.substring(0, baseline.indexOf('/') + 1) + "a".repeat(256), repeatedPath(31, 1) + "/$baseline",
                baseline + "/" + asciiPath(1025 - baseline.toByteArray(Charsets.UTF_8).size - 1), baseline + "/" + "a".repeat(4097 - baseline.length - 1), baseline + '\uD800'
            )
            val expected = listOf(NarInstallError.INVALID_PATH, NarInstallError.INVALID_PATH, NarInstallError.INVALID_PATH, NarInstallError.INVALID_PATH, NarInstallError.INVALID_PATH, NarInstallError.COMPONENT_LENGTH_LIMIT, NarInstallError.PATH_DEPTH_LIMIT, NarInstallError.PATH_LENGTH_LIMIT, NarInstallError.RAW_NAME_LENGTH_LIMIT, NarInstallError.INVALID_PATH)
            assertError(expected[index % mutations.size], validateWithPath(mutations[index % mutations.size]))
        }
    }

    @Test fun snapshotsCentralEntryExactlyOnceAndReturnsImmutableResult() {
        val oneShot = OneShotEntry()
        val result = NarArchiveInventoryValidator().validate(listOf(oneShot))
        assertTrue(result.isSuccess()); assertEquals(1, result.getInventory()!!.getEntries().size)
        try { (result.getInventory()!!.getEntries() as MutableList).clear(); throw AssertionError("inventory must be immutable") } catch (_: UnsupportedOperationException) { }
    }

    @Test fun entriesAreUnmodifiableFromJavaWithMultipleEntries() {
        val inventory = validate(file("install.txt", 1), file("payload", 1)).getInventory()!!
        assertEquals(2, inventory.getEntries().size)
        try { (inventory.getEntries() as MutableList<NarArchiveInventory.Entry>).clear(); throw AssertionError("entries must be Java-unmodifiable") } catch (_: UnsupportedOperationException) { }
    }

    private fun validate(vararg records: Record): NarArchiveInventoryResult = validate(records.toList())
    private fun validate(records: List<Record?>): NarArchiveInventoryResult { records.forEachIndexed { i, r -> if (r != null && r.ordinalValue < 0) r.ordinalValue = i }; return NarArchiveInventoryValidator().validate(records) }
    private fun validateWithPath(path: String) = validate(file("install.txt", 1), file(path, 1))
    private fun file(path: String, size: Long) = Record(path, false, sizeValue = size, compressedValue = if (size == 0L) 0 else size)
    private fun dir(path: String) = Record("$path/", true, sizeValue = 0, compressedValue = 0)
    private fun assertError(error: NarInstallError, result: NarArchiveInventoryResult) { assertFalse(result.isSuccess()); assertEquals(error, result.getError()); assertNull(result.getInventory()) }
    private fun repeatedPath(count: Int, componentLength: Int) = List(count) { "a".repeat(componentLength) }.joinToString("/")
    private fun asciiPath(bytes: Int): String { val letters = bytes - 4; val base = letters / 5; val remainder = letters % 5; return List(5) { "a".repeat(base + if (it < remainder) 1 else 0) }.joinToString("/") }
    private class Record(val raw: String?, val directory: Boolean, var crcValue: Long = 0, var methodValue: Int = if (directory) 0 else 8, var sizeValue: Long = 0, var compressedValue: Long = sizeValue) : NarArchiveInventoryValidator.CentralEntry {
        var ordinalValue = -1
        override fun getOrdinal() = ordinalValue; override fun getRawName() = raw; override fun isDirectory() = directory
        override fun getCrc() = crcValue; override fun getMethod() = methodValue; override fun getDeclaredSize() = sizeValue; override fun getCompressedSize() = compressedValue
    }
    private class ThrowingEntry : NarArchiveInventoryValidator.CentralEntry { override fun getOrdinal() = 0; override fun getRawName(): String = throw IllegalStateException(); override fun isDirectory() = false; override fun getCrc() = 0L; override fun getMethod() = 8; override fun getDeclaredSize() = 1L; override fun getCompressedSize() = 1L }
    private class OneShotEntry : NarArchiveInventoryValidator.CentralEntry {
        private var calls = 0; private fun <T> once(value: T): T { calls++; if (calls > 7) error("getter reread"); return value }
        override fun getOrdinal() = once(0); override fun getRawName() = once("install.txt"); override fun isDirectory() = once(false)
        override fun getCrc() = once(0L); override fun getMethod() = once(8); override fun getDeclaredSize() = once(1L); override fun getCompressedSize() = once(1L)
    }
}
